core/modulesource.go

```diff
@@ -250,6 +251,153 @@ func (src *ModuleSource) Pin() string {
+	var envFileExists bool
+	if err := dag.Select(ctx, dag.Root(), &envFileExists,
+		dagql.Selector{Field: "host"},
+		dagql.Selector{
+			Field: "directory",
+			Args: []dagql.NamedInput{
+				{Name: "path", Value: dagql.String(moduleDirPath)},
+				{Name: "include", Value: dagql.ArrayInput[dagql.String]{".env"}},
+			},
+		},
+		dagql.Selector{
+			Field: "exists",
+			Args: []dagql.NamedInput{
+				{Name: "path", Value: dagql.String(".env")},
+			},
+		},
+	); err != nil {
+		return nil, "", fmt.Errorf("failed to check for inner env file in %s: %w",
+			moduleDirPath, err)
+	}
+	if !envFileExists {
+		return nil, "", nil
+	}
+	envFilePath := path.Join(moduleDirPath, ".env")
+	var envFile *EnvFile
+	if err := dag.Select(ctx, dag.Root(), &envFile,
+		dagql.Selector{Field: "host"},
+		dagql.Selector{
+			Field: "file",
+			Args: []dagql.NamedInput{
+				{Name: "path", Value: dagql.String(envFilePath)},
+			},
+		},
+		dagql.Selector{
+			Field: "asEnvFile",
+			Args: []dagql.NamedInput{
+				{Name: "expand", Value: dagql.Opt(dagql.NewBoolean(true))},
+			},
+		},
+	); err != nil {
+		return nil, "", fmt.Errorf("failed to load inner env file in %s: %w", moduleDirPath, err)
+	}
+	return envFile, envFilePath, nil
+}
+
+func (src *ModuleSource) outerEnvFile(ctx context.Context) (*EnvFile, string, error) {
+	dag, err := CurrentDagqlServer(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+	var envFilePath dagql.String
+	if err := dag.Select(ctx, dag.Root(), &envFilePath,
+		dagql.Selector{Field: "host"},
+		dagql.Selector{Field: "findUp",
+			Args: []dagql.NamedInput{
+				{Name: "name", Value: dagql.NewString(".env")},
+			},
+		},
+	); err != nil {
+		return nil, "", fmt.Errorf("failed to find-up outer .env: %s", err.Error())
+	}
+	if envFilePath == "" {
+		return &EnvFile{}, "", nil
+	}
+	var envFile *EnvFile
+	if err := dag.Select(ctx, dag.Root(), &envFile,
+		dagql.Selector{Field: "host"},
+		dagql.Selector{Field: "file",
+			Args: []dagql.NamedInput{
+				{Name: "path", Value: envFilePath},
+			},
+		},
+		dagql.Selector{Field: "asEnvFile",
+			Args: []dagql.NamedInput{
+				{Name: "expand", Value: dagql.Opt(dagql.NewBoolean(true))},
+			},
+		},
+	); err != nil {
+		return nil, envFilePath.String(), fmt.Errorf("failed to load outer env file from %q: %s", envFilePath.String(), err.Error())
+	}
+	return envFile, envFilePath.String(), nil
+}
+
+// loadocalDefaults loads and merges environment files for local module defaults.
+// It combines the inner .env file (from the module's source root) with relevant
+// entries from the outer .env file (found via find-up from the host), filtered
+// by the module name and original module name as prefixes.
+//
+// Example:
+// Inner .env (mymodule/.env): FOO=bar
+// Outer .env (found via find-up): MYMODULE_BAZ=qux, OTHER_VAR=ignored
+// Result: FOO=bar, BAZ=qux (prefix "MYMODULE_" removed from outer entries)
+func (src *ModuleSource) LoadLocalDefaults(ctx context.Context) error {
```

Comment:
nit: this could maybe be refactored into a `sync.Once`'d lazy-initialized method, but I see it's really only used in various bootstraping constructors, of which there will probably only be a few, so no big deal tbh.

Reply:
I'm happy to do it, OK if I make it a follow-up PR?